Library/Homebrew/extend/os/linux/dependency_collector.rb

```diff
@@ -0,0 +1,73 @@
+# typed: true
+# frozen_string_literal: true
+
+require "os/linux/glibc"
+
+class DependencyCollector
+  extend T::Sig
+
+  undef gcc_dep_if_needed
+  undef glibc_dep_if_needed
+  undef init_global_dep_tree_if_needed!
+
+  sig { params(related_formula_names: T::Set[String]).returns(T.nilable(Dependency)) }
+  def gcc_dep_if_needed(related_formula_names)
+    return unless DevelopmentTools.system_gcc_too_old?
+    return if related_formula_names.include?(GCC)
+    return if global_dep_tree[GCC]&.intersect?(related_formula_names)
+    return if global_dep_tree[GLIBC]&.intersect?(related_formula_names) # gcc depends on glibc
+
+    Dependency.new(GCC)
+  end
+
+  sig { params(related_formula_names: T::Set[String]).returns(T.nilable(Dependency)) }
+  def glibc_dep_if_needed(related_formula_names)
+    return unless OS::Linux::Glibc.below_ci_version?
+    return if global_dep_tree[GLIBC]&.intersect?(related_formula_names)
+
+    Dependency.new(GLIBC)
+  end
+
+  private
+
+  GLIBC = "glibc"
+  GCC = CompilerSelector.preferred_gcc.freeze
+
+  # Use class variables to avoid this expensive logic needing to be done more
+  # than once.
+  # rubocop:disable Style/ClassVars
+  @@global_dep_tree = {}
+
+  sig { void }
+  def init_global_dep_tree_if_needed!
+    return unless DevelopmentTools.build_system_too_old?
+    return if @@global_dep_tree.present?
+
+    # Defined in precedence order (gcc depends on glibc).
+    global_deps = [GLIBC, GCC].freeze
+
+    @@global_dep_tree = global_deps.to_h { |name| [name, Set.new([name])] }
+
+    global_deps.each do |global_dep_name|
+      maximum_tree_depth = 10
```

Comment:
A comment explaining why `10` is appropriate here would be useful.

Reply:
@carlocab just to make sure it's good enough reasoning for you: current max depth we need to go is 8 and want to ensure that this doesn't just loop infinitely.

Reply:
That works, but a brief explanation of how we know it's `8` right now would be good.